### ISSUE TITLE
tsc: watch mode doesn't discover new files

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/TypeScript#4553]
- tsconfig with watch: true
- start tsc
- add a new file

Observe: it is not transpiled.

**Comment by mhegazy** [comment]
looks like another place where we should use the new node watch APIs: see #4643

**Comment by alexeagle** [comment]
+1 to have this soon. Is it for 1.7 like in #5076 or for 1.8 like here?

**Comment by zhengbli** [comment]
@alexeagle these two issues are both for 1.7

**Comment by dsebastien** [comment]
+1 to have this soon(er), this is annoying :)

**Comment by mhegazy** [comment]
@dbaeumer can you give latest a try?

**Comment by jmls** [comment]
Hmm - I'm using 2.0.10 and tsc -w does not spot new files. If I ctrl-c and restart, it then transpiles

**Comment by zhengbli** [comment]
@jmls Can you provide the following info:
- your os
- your node version
- are you using a network mounted file system?
- if possible, a small project that repros the issue.

**Comment by jmls** [comment]
Of course.

node: **4.6.1, 7.0.0 and 7.2.0**
os: **c9.io**, running **ubuntu 4.2.0-c9 #2 SMP x86_64 x86_64 x86_64 GNU/Linux** 

demo project at https://github.com/jmls/tsc-watch

I've created a new file by copying in the shell (`cp part1.ts part2.ts`) and creating a new file in the ide

**Comment by zhengbli** [comment]
Do you run `tsc -w` from project root path instead of from `src`? If so this is a known issue, as [node's file watcher doesn't support recursive watching on linux](https://nodejs.org/docs/latest/api/fs.html#fs_caveats).

**Comment by jmls** [comment]
I do - but I am amazed that tsc still uses fw.watch if it has such a fundamental problem. 
Are you looking to fix this by using a package like chokidar ?

**Comment by zhengbli** [comment]
Indeed it is not ideal and creating a file watcher for node is out of the domain of our main focus. If there are no signs in addressing this in upcoming node updates, I do think a third-party package is the most likely approach. Though we are very cautious in adding third-party dependencies (currently we have none).

**Comment by OliverUv** [comment]
This is a very relevant issue to my organization. I would like you to re-open it.

We, and I assume many others, do our building in Docker containers, with source directories mounted in volumes. This gives us the assurance that all devs use the same environment as production. The Docker container, of course, is running Linux, where node's fs.watch is broken.

We let the tsc --watch process inside Docker pick up changes made on the outside, giving us a very nice dev cycle where compilation and tests are re-run automatically whenever a file is changed.

We absolutely want to keep using the native `tsc --watch`, since external modules have many issues:

* Can never do as good a job with incremental recompilation as tsc can
* Stop working if tsc changes
* Maintainers stop maintaining them

So I ask you: Given the prevalence of people using Docker for development, please start considering Linux a supported platform, and ensure that your amazing `--watch` mode is fully supported there as well.

Since you don't want strict dependencies on external libraries, I suggest doing this with an optionalDependency.

**Comment by ORESoftware** [comment]
```
const chokidar = require('chokidar');
const watcher = chokidar.watch(<dir>);

watcher.on('link', function(p){
 /// a file was added with path = p
});
```

this should be easy to add to tsc -w, right?

**Comment by ORESoftware** [comment]
Alright so I already had created a tsc --watch wrapper that could spawn separate watchers for each tsconfig.json file in a project.

So I added this chokidar feature, so now the watcher will restart when a new file is added. I created this recently, so it's nothing great, but star it if you are interested in me making it better.

https://github.com/ORESoftware/tsc-multi-watch
<img width="1029" alt="screenshot 2017-06-19 22 20 06" src="https://user-images.githubusercontent.com/11139560/27317844-03a043ec-553e-11e7-9970-1782275b2913.png">
